lincheck/src/main/java/org/jetbrains/kotlinx/lincheck/verifier/LTS.kt

```diff
@@ -84,313 +91,377 @@ abstract class AbstractLTSVerifier<STATE>(val scenario: ExecutionScenario, val t
-        override val context: CoroutineContext
-            get() = VerifierInterceptor(ticket, actor)
+private val relaxedActors = IdentityHashMap<Method, Boolean>()
+val Actor.isRelaxed: Boolean
+    get() = relaxedActors.computeIfAbsent(this.method) {
+        this.method.isAnnotationPresent(QuantitativeRelaxed::class.java)
+    }
 
-        override fun resumeWith(result: kotlin.Result<Any?>) {
+/**
+ * When suspended operation is resumed by another thread,
+ * [VerifierInterceptor.interceptContinuation] is called to intercept it's continuation.
+ * Intercepted continuation just writes the result of the suspension point and reference to the unintercepted continuation
+ * so that the calling thread could resume this continuation by itself.
+ */
+internal class VerifierInterceptor(
+    private val ticket: Int,
+    private val actor: Actor,
+    private val resumedTicketsWithResults: MutableMap<Int, ResumedResult>
+) : AbstractCoroutineContextElement(ContinuationInterceptor),
+    ContinuationInterceptor {
+    override fun <T> interceptContinuation(continuation: Continuation<T>): Continuation<T> {
+        return Continuation(EmptyCoroutineContext) { res ->
             resumedTicketsWithResults[ticket] =
-                ResumedResult(null as Continuation<Any?>? to result).also { it.resumedActor = actor }
+                ResumedResult(continuation as Continuation<Any?> to res).also { it.resumedActor = actor }
         }
     }
+}
 
+/**
+ * Passed as an argument to invoke suspendable operation
+ *
+ * [context] contains `ContinuationInterceptor` that intercepts continuation of the resumed operation
+ * separating `request` and `follow-up` parts.
+ *
+ * If the operation has `follow-up` phase then [resumeWith] is executed at the end of the `follow-up`
+ * writing the final result of the operation. Otherwise if there is no `follow-up`,
+ * meaning that result of the suspension point equals the final result of the operation,
+ * [resumeWith] will be executed just after resumption by the resuming thread.
+ */
+internal class Completion(
+    private val ticket: Int,
+    private val actor: Actor,
+    private val resumedTicketsWithResults: MutableMap<Int, ResumedResult>
+) : Continuation<Any?> {
+    override val context: CoroutineContext
+        get() = VerifierInterceptor(ticket, actor, resumedTicketsWithResults)
+
+    override fun resumeWith(result: kotlin.Result<Any?>) {
+        resumedTicketsWithResults[ticket] =
+            ResumedResult(null as Continuation<Any?>? to result).also { it.resumedActor = actor }
+    }
+}
+
+/**
+ * `Operation` is divided into the `request` part that can either suspend it's execution
+ * or return the final result immediately
+ * and `follow-up` part that follows the `request` when the `Operation` is resumed.
+ *
+ * [ticket] is assigned to the suspended request and is passed to the corresponding follow-up.
+ */
+data class Operation(
+    val actor: Actor,
+    val expectedResult: Result,
```

Comment:
We have to store the `expectedResult` to pass it at the invocation of cost counter methods in the relaxed mode.